rfc/0077-zone-aware-replica-reads.md

```diff
@@ -0,0 +1,183 @@
+# Meta
+
+| Field          | Value                    |
+|----------------|--------------------------|
+| RFC Name       | Zone-Aware Replica Reads |
+| RFC ID         | 77                       |
+| Start Date     | 2024-04-25               |
+| Owner          | Sergey Avseyev           |
+| Current Status | DRAFT                    |
+| Revision       | #1                       |
+
+# Summary
+
+This document describes changes that SDKs must implement to optimize costs of
+network traffic when the server is being deployed across multiple availability
+zones.
+
+# Motivation
+
+Existing SDKs do not take into account node availability groups of the nodes in
+the cluster. In particular, when performing `getAllReplicas` operations, the SDK
+would send `GET (0x00)` request to active vBucket and `GET_REPLICA (0x83)` to
+each of the replicas. This approach will make reading from replicas expensive
+operations if nodes of the cluster deployed in different availability zones
+(AZ).
+
+# Use Cases
+
+## No Preference
+
+This is current behaviour, where the SDK does not take into account server
+groups.
+
+![figure 1: No Prefrence](figures/0077-case-1-no-preference.svg)
+
+## Local Only
+
+In this case, all reads will be done only from the local availability zone. It
+is the cheapest solution, although it reduces chances of getting data from
+replicas.
+
+![figure 2: Local Only](figures/0077-case-2-local-only.svg)
+
+## Local First
+
+This case is a compromise between cost and availability, which trades speed. The
+idea is that the SDK will partition set of `[active, replica1, ...]` nodes into
+two groups: local and non-local. Then it will perform requests to local only,
+and only when neither of nodes will return data, send requests to non-local
+group.
+
+![figure 3: Local First](figures/0077-case-3-local-first.svg)
+
+# Changes
+
+New enumeration should be defined for expressing different strategies:
+
+```
+enum ReadPreference {
+    NO_PREFERENCE,
+    LOCAL_ONLY,
+    LOCAL_FIRST,
+};
+```
+
+Each of the operation should get new option, with default value of
+`ReadPreference::NO_PREFERENCE`.
+
+```
+class GetAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class GetAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+In all failure cases, when the SDK cannot handle operation, it must return `102
+DocumentUnretrievable` with human-readable explanation whenever it is possible
```

Comment:
One reason I was avoiding using the existing Get[Any|All]Replicas, is I consider their interfaces rather broken.  One concrete example - I have no easy way of disambiguating programmatically between the document not existing, and a network error.  There are other issues, including that GetAllReplicas swallows all substream errors, making it not a good building block for higher-level replica operations.

We need to consider that the feature needs to be added to transactions also - in fact this is the most important use-case since it is the one the requesting customer will use.  And I'm somewhat reluctant to add a broken interface into transactions.

Reply:
On transactions - let's focus on non-transactions for now though.  Once we nail that API it should be relatively straightforward to add it to transactions too.  I'm only bringing up transactions as an argument for why I'd maybe prefer we avoided using getAnyReplica() for this.

Reply:
new options are essentially change behavior, maybe we can agree that with this new option it will distinguish 
* network error 
* document is not found
* case where the selected strategy does not allow us to fulfill request
  * not enough local replicas
  * no replicas configured
  * even master is not local group

Reply:
Yes, that seems a viable solution.

Reply:
I may be changing my feeling on that..  I'm very reluctant to add a not-great interface into transactions.  

It is an option to only include getAnyReplica() in transactions, and only allow LOCAL_ONLY mode (with its better behaviour) - but perhaps not a great one.

Reply:
I believe after all it makes sense to add getAllReplicas and getAnyReplica into transactions and let them behave like regular operations, although implemented using lookupInAllReplicas and lookupInAnyReplica internally.